src/protocol/sort/shuffle.rs

```diff
@@ -156,21 +158,21 @@ impl<'a, F: Field> Shuffle<'a, F> {
         &mut self,
         ctx: ProtocolContext<'_, Replicated<F>, F>,
         permutations: &mut (Permutation, Permutation),
-    ) -> Result<(), BoxError>
+    ) -> Result<Vec<Replicated<F>>, BoxError>
     where
         F: Field,
     {
-        *self.input = self
+        self.input = self
             .shuffle_or_unshuffle_once(ShuffleOrUnshuffle::Shuffle, &ctx, Step1, permutations)
             .await?;
-        *self.input = self
+        self.input = self
             .shuffle_or_unshuffle_once(ShuffleOrUnshuffle::Shuffle, &ctx, Step2, permutations)
             .await?;
-        *self.input = self
+        self.input = self
             .shuffle_or_unshuffle_once(ShuffleOrUnshuffle::Shuffle, &ctx, Step3, permutations)
             .await?;
 
-        Ok(())
+        Ok(take(&mut self.input))
```

Comment:
I'd suggest doing the same thing I suggested above - just eliminate the struct and member variables, and make `shuffle_permutation` just a function that takes ownership over the vector you pass in.

Reply:
Shuffle is doing a lot more than just taking this as an input. `shuffle_or_unshuffle_once` modifies the input and needs a mut ref for it. Have changed the other ones for now.